ex03.exs

```diff
@@ -51,14 +51,77 @@ defmodule Ex03 do
   """
 
   def pmap(collection, process_count, function) do
-    # your code goes here
+    collection
+      |> splitIntoChunks(process_count)
+      |> convertToProcesses(function)
+      |> runProcesses()
+
     # I'm hoping to see a simple pipeline as the body of this function...
   end
 
-  # and here...
+  def splitIntoChunks(collection, numProcesses) do
+    countElements = Enum.count(collection) / numProcesses
+      |> Kernel.trunc()
+    Enum.chunk_every(collection, countElements, countElements,[])
+      |> ensureCorrectNumChunks(numProcesses)
+  end
+
+  def ensureCorrectNumChunks(collection, numProcesses) do
+    if length(collection) != numProcesses do
+      {leftover, _restCollection} = List.pop_at(collection, numProcesses - 1)
+      List.update_at(collection, numProcesses - 1, &(List.insert_at(&1, numProcesses, leftover)))
+    end
+    collection
+  end
+
+  def convertToProcesses(collection, function) do
+    Enum.map(
+      collection,
+      fn subCollection -> getProcessForSublist(subCollection, function) end
+    )
+  end
+
+  def getProcessForSublist(subCollection, function) do
+    spawn fn () -> applyGivenFunction(function, subCollection) end
+  end
+
+  def applyGivenFunction(function, list) do
+    receive do
+      {:apply, from} ->
+        newList = Enum.map(list, fn el -> function.(el) end)
+        send from, {:new_list, newList}
+        applyGivenFunction(function, newList)
+    end
+  end
+
+  def runProcesses(collection) do
+    Enum.flat_map(collection, fn pid -> runProcess(pid) end)
+  end
+
+  def runProcess(pid) do
+    send pid, {:apply, self()}
+    receive do
+      {:new_list, newList} ->
+        send pid, newList
+    end
+
+  end
 
 end
```

Comment:
I'll show a solution in class, but you were on the right track. You just missed out on the idea that the process should be passed the function when spawned, so it can start running asynchronously.

# -25 as per rubric